### PR TITLE
detect docker btrfs backend filesystem

### PR DESCRIPTION
**Pull request by aojea**
Fixes: #1945,#2050

### CODE REVIEW

**Comment by aojea** [comment]
/assign @BenTheElder 
@anfredette can you please verify this solves the issue for fedora33?

File: pkg/cluster/internal/providers/docker/util.go [hunk 1]
```diff
@@ -48,15 +48,29 @@ func usernsRemap() bool {
...
+	// docker info -f '{{json .DriverStatus  }}'
+	// [["Backing Filesystem","extfs"],["Supports d_type","true"],["Native Overlay Diff","true"]]
+	cmd = exec.Command("docker", "info", "-f", "{{ index (index .DriverStatus 0) 1 }}")
+	lines, err = exec.OutputLines(cmd)
+	if err != nil || len(lines) != 1 {
+		return false
 	}
-	return storage == "btrfs" || storage == "zfs"
+	storage = strings.ToLower(strings.TrimSpace(lines[0]))
+	return storage == "btrfs"
```
Comment (BenTheElder) [inline pkg/cluster/internal/providers/docker/util.go:75]: We also should do this for xfs, zfs.

Comment (aojea) [reply pkg/cluster/internal/providers/docker/util.go:75]: in the driver too?, we weren´t doing it for xfs

File: pkg/cluster/internal/providers/docker/util.go [hunk 2]
```diff
@@ -48,15 +48,29 @@ func usernsRemap() bool {
...
 	lines, err := exec.OutputLines(cmd)
-	if err != nil {
+	if err != nil || len(lines) != 1 {
 		return false
 	}
-	if len(lines) > 0 {
-		storage = strings.ToLower(strings.TrimSpace(lines[0]))
+
+	storage = strings.ToLower(strings.TrimSpace(lines[0]))
+	if storage == "btrfs" || storage == "zfs" {
```
Comment (aojea) [inline pkg/cluster/internal/providers/docker/util.go:62]: do I need to change this too @BenTheElder ?

**Review by anfredette — APPROVED** [review]
I'm running on a fresh install of Fedora 33 Workstation in which btrfs replaced ext4 as the default filesystem.

kind was failing to start with the following error:
ERROR: failed to create cluster: failed to init node with kubeadm: command "docker exec --privileged ovn-control-plane kubeadm init --skip-phases=preflight --config=/kind/kubeadm.conf --skip-token-print --v=6" failed with error: exit status 1

For reference:
$ docker info --format='{{ index (index .DriverStatus 0) 1 }}'
btrfs

This change fixes my issue.

File: pkg/cluster/internal/providers/docker/util.go (hunk 2 above)
Comment (BenTheElder) [reply pkg/cluster/internal/providers/docker/util.go:62]: we should add `devicemapper` here https://github.com/kubernetes-sigs/kind/issues/2050#issue-800750520

Comment (aojea) [reply pkg/cluster/internal/providers/docker/util.go]: done

File: pkg/cluster/internal/providers/docker/util.go (hunk 1 above)
Comment (aojea) [reply pkg/cluster/internal/providers/docker/util.go]: done

File: pkg/cluster/internal/providers/docker/util.go [hunk 3]
```diff
@@ -48,15 +48,29 @@ func usernsRemap() bool {
...
+
+	storage = strings.ToLower(strings.TrimSpace(lines[0]))
+	if storage == "btrfs" || storage == "zfs" || storage == "devicemapper" {
+		return true
+	}
+
+	// check the backing file system
+	// docker info -f '{{json .DriverStatus  }}'
+	// [["Backing Filesystem","extfs"],["Supports d_type","true"],["Native Overlay Diff","true"]]
+	cmd = exec.Command("docker", "info", "-f", "{{ index (index .DriverStatus 0) 1 }}")
```
Comment (BenTheElder) [inline pkg/cluster/internal/providers/docker/util.go:69]: do we have reason to believe this is always index 0 of DriverStatus?

Comment (aojea) [reply pkg/cluster/internal/providers/docker/util.go:69]: I tested it in 3 different deployments, maybe @AkihiroSuda can help us here ^^^

Comment (BenTheElder) [reply pkg/cluster/internal/providers/docker/util.go:69]: This seems unnecessarily fragile we can decode the structure and inspect the tuple with "Backing Filesystem"

File: pkg/cluster/internal/providers/docker/util.go [hunk 4]
```diff
@@ -48,15 +49,39 @@ func usernsRemap() bool {
...
+	if err != nil || len(lines) != 1 {
+		return false
+	}
+	var dat [][]string
+	if err := json.Unmarshal([]byte(lines[0]), &dat); err != nil {
+		return false
+	}
+	for _, item := range dat {
+		if item[0] == "Backing Filesystem" {
+			storage = strings.ToLower(strings.TrimSpace(item[1]))
```
Comment (BenTheElder) [inline pkg/cluster/internal/providers/docker/util.go:81]: i'm not sure we should be trimspace etc. here, but that can be tweaked in a follow-up

Comment (aojea) [reply pkg/cluster/internal/providers/docker/util.go:81]: let's do it now

**Comment by BenTheElder** [comment]
/lgtm
/approve
thanks!

**Pull request merged**